### PR TITLE
Replace kyverno rolebinding policies with cloud agent

### PR DESCRIPTION
**Pull request by glrf**
## Checklist

- [x] PR contains a single logical change (to build a better changelog).
- [x] Update the documentation.
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog.
- [ ] Link this PR to related issues or PRs.

### CODE REVIEW

File: docs/modules/ROOT/pages/explanations/policies.adoc [hunk 1]
```diff
@@ -231,13 +231,16 @@ To protect the cluster, we need to ensure that users can't modify arbitrary labe
...
 TIP: This policy is executed when namespaces are created or updated.
 
-xref:references/policies/10_generate_default_rolebinding_in_ns.adoc[`default-rolebinding-in-ns` (labeled `ORG_RBAC`)]::
-This policy is applied to any namespace which is created with an organization label.
-It grants role `admin` to the organization to which the new namespace belongs.
+`default-rolebinding-in-ns (labeled ORG_RBAC)`::
+The https://github.com/appuio/appuio-cloud-agent/blob/master/controllers/org_rbac_controller.go[Organization RBAC Controller] of the APPUiO Cloud Agent makes sure that by default organizations are granted all the permissions necessary to manager their own namespaces.
++
+It does this by creating `RoleBindings` in every organization namespace that grant configured `ClusterRoles`.
+Most importantly it grants role `admin` to the organization to which the new namespace belongs.
```
Comment (simu) [inline docs/modules/ROOT/pages/explanations/policies.adoc:238]:
```suggestion
Most importantly, it grants role `admin` to the organization to which the new namespace belongs.
```

File: docs/modules/ROOT/pages/explanations/policies.adoc [hunk 2]
```diff
@@ -231,13 +231,16 @@ To protect the cluster, we need to ensure that users can't modify arbitrary labe
 +
 TIP: This policy is executed when namespaces are created or updated.
 
-xref:references/policies/10_generate_default_rolebinding_in_ns.adoc[`default-rolebinding-in-ns` (labeled `ORG_RBAC`)]::
-This policy is applied to any namespace which is created with an organization label.
-It grants role `admin` to the organization to which the new namespace belongs.
+`default-rolebinding-in-ns (labeled ORG_RBAC)`::
+The https://github.com/appuio/appuio-cloud-agent/blob/master/controllers/org_rbac_controller.go[Organization RBAC Controller] of the APPUiO Cloud Agent makes sure that by default organizations are granted all the permissions necessary to manager their own namespaces.
++
+It does this by creating `RoleBindings` in every organization namespace that grant configured `ClusterRoles`.
```
Comment (simu) [inline docs/modules/ROOT/pages/explanations/policies.adoc:237]:
```suggestion
It does so by creating `RoleBindings` in every organization namespace that grant configured `ClusterRoles`.
```

File: docs/modules/ROOT/pages/explanations/policies.adoc [hunk 3]
```diff
@@ -231,13 +231,16 @@ To protect the cluster, we need to ensure that users can't modify arbitrary labe
...
+Most importantly it grants role `admin` to the organization to which the new namespace belongs.
 +
-This fulfils the requirement that https://kb.vshn.ch/appuio-cloud/references/quality-requirements/usability/ns-organization-rbac.html[namespaces are owned by organizations].
+The controller will only make sure that the `RoleBindings` exit and won't modify existing `RoleBindings`.
```
Comment (simu) [inline docs/modules/ROOT/pages/explanations/policies.adoc:240]:
```suggestion
The controller will only make sure that the `RoleBindings` exist and won't modify existing `RoleBindings`.
```

File: component/agent.jsonnet [hunk 4]
```diff
@@ -165,6 +187,23 @@ local metricsService = loadManifest('manager/service.yaml') {
       },
     ],
   },
+  '01_default_org_role_binding': [
+    kube.ClusterRoleBinding('appuio-cloud-agent:%s' % [ cr ]) {
```
Comment (simu) [inline component/agent.jsonnet:191]: Would be nice to add a comment here why this is necessary (the agent needs all permissions it wants to grant in new RoleBindings).

**Review by simu — COMMENTED** [review]
LGTM overall.

I have a broad question/comment reqarding the namespace ownership clusterrole name:

IIRC we provided the parameter `generatedNamespaceOwnerClusterRole.name` so that users of the component can customize the name of the generated clusterrole and rolebinding. This isn't really possible anymore, since the new implementation hard-codes the clusterrole name to `namespace-owner` unless the deprecated parameter is set. Nontheless, we add the rolebinding in the default agent config, but there's no validation logic to prevent accidental misconfigurations.

I think I'd prefer either

* having a parameter which controls the name of the namespace owner clusterrole and reference that parameter in the default agent config via reclass reference.
* dropping the namespace-ownership config from the agent default config and injecting it in the component using the hard-coded clusterrole name.

I'm not sure which of these options I like better, but both of them are more consistent than the current approach in the PR.

**Comment by glrf** [comment]
> I have a broad question/comment reqarding the namespace ownership clusterrole name:
> ...

That's a good input and you're right the current solution isn't very consistent.

> * having a parameter which controls the name of the namespace owner clusterrole and reference that parameter in the default agent config via reclass reference.
> * dropping the namespace-ownership config from the agent default config and injecting it in the component using the hard-coded clusterrole name.

A third option could be to make this very generic: allow deploying arbitrary clusterroles and add a default namespace owner clusterrole which is also referenced in the agent default config. But I'm not sure if this really makes sense or if it's too generic.

**Comment by simu** [comment]
> A third option could be to make this very generic: allow deploying arbitrary clusterroles and add a default namespace owner clusterrole which is also referenced in the agent default config. But I'm not sure if this really makes sense or if it's too generic.

Thinking about it again, this approach seems to make the most sense, as it would allow:
* adding new roles (think e.g. user workload monitoring rolebindings) by only updating the component defaults
* testing changes purely through the hierarchy
* the possibility of deploying different sets of roles on different cloud providers (e.g. GPU access or similar)

Additionally the approach is also in line with other components which have support for semi-generic resource creation (secrets and similar)

**Comment by glrf** [comment]
I switched to the more generic way to generate clusterroles. Staying backwards compatible was bit of a PITA.

**Review by simu — APPROVED** [review]
LGTM now.

I think we should start mapping out a road to component appuio-cloud v2 (probably after the migration from Kyverno to agent policies is complete). That way we'll have a target point at which we can drop all the support for deprecated configs.

**Pull request merged**